wandb/sdk/service/server.py

```diff
@@ -586,51 +343,59 @@ def ServerInformTeardown(  # noqa: N802
         return result
 
 
-class GrpcServer:
+class WandbServer:
+    _pid: Optional[int]
+    _grpc_port: Optional[int]
+    _sock_port: Optional[int]
+    _debug: bool
+    _serve_grpc: bool
+    _serve_sock: bool
+
     def __init__(
         self,
-        port: int = None,
+        grpc_port: int = None,
+        sock_port: int = None,
         port_fname: str = None,
         address: str = None,
         pid: int = None,
         debug: bool = False,
+        serve_grpc: bool = False,
+        serve_sock: bool = False,
     ) -> None:
-        self._port = port
+        self._grpc_port = grpc_port
+        self._sock_port = sock_port
         self._port_fname = port_fname
         self._address = address
         self._pid = pid
         self._debug = debug
+        self._serve_grpc = serve_grpc
+        self._serve_sock = serve_sock
         debug = True
         if debug:
             logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
 
-    def _inform_used_port(self, port: int) -> None:
+    def _inform_used_ports(
+        self, grpc_port: Optional[int], sock_port: Optional[int]
+    ) -> None:
         if not self._port_fname:
             return
-        dname, bname = os.path.split(self._port_fname)
-        f = tempfile.NamedTemporaryFile(prefix=bname, dir=dname, mode="w", delete=False)
-        tmp_filename = f.name
-        try:
-            with f:
-                f.write("%d" % port)
-            os.rename(tmp_filename, self._port_fname)
-        except Exception:
-            os.unlink(tmp_filename)
-            raise
+        pf = port_file.PortFile(grpc_port=grpc_port, sock_port=sock_port)
+        pf.write(self._port_fname)
 
-    def _launch(self, mux: StreamMux) -> int:
+    def _start_grpc(self, mux: StreamMux) -> int:
         address: str = self._address or "127.0.0.1"
-        port: int = self._port or 0
+        port: int = self._grpc_port or 0
         pid: int = self._pid or 0
```

Comment:
should we just define the default values in `__init__` why use `None`?

Reply:
I rarely like default args as then you cant tell the difference between one passed or not... so i tend to (overly) avoid them, but they make sense in some cases --  I will look at this one and see.

Reply:
I want to keep it as it is.   _grpc_port of None is very different than _grpc_port of 0. 
0 means that the port should be determined by the server.  None means the port isnt used.